osdep/threads-win32.h

```diff
@@ -0,0 +1,222 @@
+    (void) cond;
+    return 0;
+}
+
+static inline int mp_cond_broadcast(mp_cond *cond)
+{
+    WakeAllConditionVariable(cond);
+    return 0;
+}
+
+static inline int mp_cond_signal(mp_cond *cond)
+{
+    WakeConditionVariable(cond);
+    return 0;
+}
+
+static inline int mp_cond_timedwait(mp_cond *cond, mp_mutex *mutex, int64_t timeout)
+{
+    timeout = MPCLAMP(timeout, 0, MP_TIME_MS_TO_NS(INFINITE)) / MP_TIME_MS_TO_NS(1);
+
+    int ret = 0;
+    int hrt = mp_start_hires_timers(timeout);
+    BOOL bRet;
+
+    if (mutex->use_cs) {
+        bRet = SleepConditionVariableCS(cond, &mutex->cs, timeout);
+    } else {
+        bRet = SleepConditionVariableSRW(cond, &mutex->srw, timeout, 0);
+    }
+    if (bRet == FALSE)
+        ret = GetLastError() == ERROR_TIMEOUT ? ETIMEDOUT : EINVAL;
+
+    mp_end_hires_timers(hrt);
+    return ret;
+}
+
+static inline int mp_cond_wait(mp_cond *cond, mp_mutex *mutex)
+{
+    return mp_cond_timedwait(cond, mutex, MP_TIME_MS_TO_NS(INFINITE));
+}
+
+static inline int mp_cond_timedwait_until(mp_cond *cond, mp_mutex *mutex, int64_t until)
+{
+    return mp_cond_timedwait(cond, mutex, until - mp_time_ns());
+}
+
+static inline int mp_exec_once(mp_once *once, void (*init_routine)(void))
+{
+    BOOL pending;
+
+    if (!InitOnceBeginInitialize(once, 0, &pending, NULL))
+        abort();
+
+    if (pending) {
+        init_routine();
+        InitOnceComplete(once, 0, NULL);
+    }
+
+    return 0;
+}
+
+#define MP_THREAD_VOID unsigned __stdcall
+#define MP_THREAD_RETURN() return 0
+
+static inline int mp_thread_create(mp_thread *thread,
+                                   MP_THREAD_VOID (*fun)(void *),
+                                   void *__restrict arg)
+{
+    *thread = (HANDLE) _beginthreadex(NULL, 0, fun, arg, 0, NULL);
+    return *thread ? 0 : -1;
+}
+
+static inline int mp_thread_join(mp_thread thread)
+{
+    DWORD ret = WaitForSingleObject(thread, INFINITE);
+    if (ret != WAIT_OBJECT_0)
+        return ret == WAIT_ABANDONED ? EINVAL : EDEADLK;
+    CloseHandle(thread);
+    return 0;
+}
+
+static inline int mp_thread_join_id(mp_thread_id id)
+{
+    mp_thread thread = OpenThread(SYNCHRONIZE, FALSE, id);
+    if (!thread)
+        return ESRCH;
+    int ret = mp_thread_join(thread);
+    if (ret)
+        CloseHandle(thread);
+    return ret;
+}
+
+static inline int mp_thread_detach(mp_thread thread)
+{
+    return CloseHandle(thread) ? 0 : EINVAL;
+}
+
+#define mp_thread_current_id GetCurrentThreadId
+#define mp_thread_id_equal(a, b) ((a) == (b))
+#define mp_thread_get_id(thread) GetThreadId(thread)
+
+wchar_t *mp_from_utf8(void *talloc_ctx, const char *s);
+static inline void mp_thread_set_name(const char *name)
+{
+#if !HAVE_UWP
+    HMODULE kernel32 = GetModuleHandleW(L"kernel32.dll");
+    if (!kernel32)
+        return;
+    HRESULT (WINAPI *pSetThreadDescription)(HANDLE, PCWSTR) =
+        (void *) GetProcAddress(kernel32, "SetThreadDescription");
+    if (!pSetThreadDescription)
```

Comment:
[docs](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setthreaddescription) say:
> Windows Server 2016, Windows 10 LTSB 2016 and Windows 10 version 1607: SetThreadDescription is only available by Run Time Dynamic Linking in KernelBase.dll.

is that why we're doing this?

BTW: the docs also say that is function _is_ available in UWP :thinking:

Reply:
I kept it as it was with dynamic loading the symbol. We still support Windows 8. I guess UWP branch could call it directly.